Remove multi-map delete API

jordalgo opened the pull request:
This is no longer valid:
```
@a = 1;
@b[1] = 2;

delete(@a, @b[1]);
```

This also fixes map key resizing for the `delete` function: this is no longer an error: `@y["hi"] = 5; delete(@y, "longerstr");`

https://github.com/bpftrace/bpftrace/issues/3495
https://github.com/bpftrace/bpftrace/issues/3496

<!--
Please provide a description of your change below this comment.

Then please complete the checklist.

Useful contribution guidelines and tips are in docs/developers.md.

Warning: please make sure that you have implemented and tested your
         change against the latest version of bpftrace (unless opening a
         PR for a release branch).
-->

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

jordalgo commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -754,18 +754,19 @@ void CodegenLLVM::visit(Call &call)
-      }
+    auto &arg0 = *call.vargs.at(0);
+    auto &map = static_cast<Map &>(arg0);
+    auto [key, scoped_key_deleter] = call.vargs.size() > 1
+                                         ? getMapKey(map, call.vargs.at(1))
+                                         : getMapKey(map);
```
Cool. Will add.

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -549,7 +549,7 @@ namespace {
 {
   return call.func == "print" || call.func == "clear" || call.func == "zero" ||
-         call.func == "len" || call.func == "has_key";
+         call.func == "len" || call.func == "has_key" || call.func == "delete";
```
Good catch. Fixed in the latest.

ajor commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -756,18 +756,22 @@ void CodegenLLVM::visit(Call &call)
-      }
+    auto &arg0 = *call.vargs.at(0);
+    auto &map = static_cast<Map &>(arg0);
+    // Note: delete accepts two arguments except in the case of scalar maps
+    // (maps with no keys) in which case it you can just pass it the map
+    // and it will act similar to `clear` e.g. `delete(@scalar);`
```
Also supports one argument for the legacy delete syntax!
`delete(@map[123])`

jordalgo replied:
Haha fine. But that's going away.

ajor replied:
I'm not sure it should go away - it's been part of the language for so long that all existing scripts are using it. Should we really break them?

jordalgo replied:
We can create a separate issue to discuss but my 2cents - yes we should deprecate and eventually remove. From someone else reading the script the syntax makes no sense. `delete(@map[123])` looks like we're deleting the value and deleting it from what? Also do we want to set a precedence that we can't delete anything from the bpftrace API? We could do this in phases e.g. begin by one release issuing a warning then the next release it gets deleted.

jordalgo replied:
Either way, I updated the comment language in the codegen file.

jordalgo commented:
> Sorry, found another case that shouldn't be accepted:

Don't be sorry! I wish I had thought of these cases. Anyway, fixed the latest one and added a test case.

ajor commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -724,46 +724,48 @@ void SemanticAnalyser::visit(Call &call)
+          auto *map_key_type = get_map_key_type(map);
+          if (map_key_type) {
+            auto &arg1 = *call.vargs.at(1);
+            if (is_final_pass() && arg1.is_map) {
+              Map &map_as_key = static_cast<Map &>(arg1);
+              auto *map_as_key_key_type = get_map_key_type(map_as_key);
+              if (!map_as_key.key_expr && map_as_key_key_type &&
+                  !map_as_key_key_type->IsNoneTy()) {
+                LOG(ERROR, map_as_key.loc, err_)
+                    << "Argument mismatch for " << map_as_key.ident << ": "
+                    << "trying to access with no arguments when map expects "
+                       "arguments: "
+                       "'"
+                    << *map_as_key_key_type << "'";
```
It looks like we're duplicating some code from `visit(Map &map)` here. It might be cleaner to keep all key validation logic in that function.

The issue seems to be that we're setting `map.skip_key_validation` for delete when we don't actually want to skip validation for the second argument. What about changing the call to `skip_key_validation(const Call &call)` to only run on the first argument? https://github.com/bpftrace/bpftrace/blob/84cfa952f5c4d34e6b7771fdf42e1b7dd422b011/src/ast/passes/semantic_analyser.cpp#L592

jordalgo replied:
Good call. This made me realize I also need to apply this same logic for `has_key`.

ajor reviewed (approved):
Nice!
